eval/evaluation.py

```diff
@@ -0,0 +1,330 @@
+        collection (List(dict)): collection of conversations.
+
+    Returns:
+        List[List]: List of utterances in a list of conversations
+    """
+    return [[utterance['text']
+             for utterance in conversation]
+            for conversation in collection]
+
+
+def get_segments(collection, annotationTypes=None):
+    """Given annotation types return all semantic annotations of that type for
+    all utterances.
+
+    Args:
+        collection (List(dict)): List of conversations.
+        annotationTypes (dict, optional): Dictionary with annotation tye -
+        entity type pairs. Defaults to None.
+
+    Returns:
+        List(dict): Collection with annotations equal to annotation types.
+    """
+    segments = []
+    for conversation in collection:
+        conversation_segment = []
+        for utterance in conversation:
+            if not annotationTypes:
+                conversation_segment.append(utterance['segments'])
+                continue
+
+            segments_to_keep = []
+            for segment in utterance['segments']:
+                annotations = [
+                    ann for ann in segment['annotations']
+                    if ann['entityType'] in annotationTypes.get(
+                        ann['annotationType'], [])
+                ]
+
+                if len(annotations) > 0:
+                    to_keep = deepcopy(segment)
+                    to_keep['annotations'] = annotations
+                    segments_to_keep.append(to_keep)
+            conversation_segment.append(segments_to_keep)
+        segments.append(conversation_segment)
+    return segments
+
+
+def get_annotation_types(collection):
+    """Get all annotation types that exist in the evaluation set.
+
+    Args:
+        collection (List(Dict)): Collection of conversations.
+
+    Returns:
+        Dict: All annotation types.
+    """
+    annotation_types = defaultdict(set)
+    for utterance in collection:
+        for segment in utterance['segments']:
+            for annotation in segment['annotations']:
+                annotation_types[annotation['annotationType']].add(
+                    annotation['entityType'])
+
+    return annotation_types
+
+
+def get_slot_annotator(config):
+    """Load an agent with a given config.
+
+    Args:
+        config (Dict): Contents of a configuration file.
+
+    Returns:
+        SlotAnnotator: Returns instance of SlotAnnotator.
+    """
+    agent = Agent(config)
+    agent.initialize()
+    return agent.nlu.intents_checker.slot_annotator
+
+
+def annotate_slot_utterance(annotator, slot, utterance):
+    """Annotates single utterance for single slot.
+
+    Args:
+        annotator (SlotAnnotator): Annotator used for annotation.
+        slot (Slots): Slot for which to annotate.
+        utterance (Text): Raw text for which to annotate.
+
+    Returns:
+        List[Dict]: List of semantic annotations for utterance.
+    """
+    user_utterance = UserUtterance({'text': utterance})
```

Comment:
It should be possible to use the context (i.e., previous utterances) when annotating a given user utterance (even if current methods don't consider context).